### PR TITLE
Converted ZC_USE_SKILL to struct

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: both

* **Description of Pull Request**: Implemented ZC_USE_SKILL packet usage

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -6088,46 +6088,36 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
-{
-	unsigned char buf[17];
-#if PACKETVER < 20130731
-	const int cmd = 0x11a;
-#else
-	const int cmd = 0x9cb;
-#endif
-	int offset = 0;
-	bool success = ( tick != 0 );
+bool clif_skill_nodamage(struct block_list* src,block_list &dst, uint16 skill_id, int heal, t_tick tick){
```
Comment (Atemo) [inline src/map/clif.cpp:6091]:
```suggestion
bool clif_skill_nodamage( block_list* src, block_list& dst, uint16 skill_id, int32 heal, bool success ){
```

File: src/map/clif.hpp [hunk 2]
```diff
@@ -940,7 +940,7 @@ void clif_skill_fail( map_session_data& sd, uint16 skill_id, enum useskill_fail_
 void clif_skill_cooldown( map_session_data &sd, uint16 skill_id, t_tick tick );
 int clif_skill_damage(struct block_list *src,struct block_list *dst,t_tick tick,int sdelay,int ddelay,int64 sdamage,int div,uint16 skill_id,uint16 skill_lv,enum e_damage_type type);
 //int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick,int sdelay,int ddelay,int damage,int div,uint16 skill_id,uint16 skill_lv,enum e_damage_type type);
-bool clif_skill_nodamage(struct block_list *src,struct block_list *dst,uint16 skill_id,int heal,t_tick tick);
+bool clif_skill_nodamage(struct block_list* src, block_list& dst,uint16 skill_id,int heal,t_tick tick);
```
Comment (Atemo) [inline src/map/clif.hpp:943]:
```suggestion
bool clif_skill_nodamage( block_list* src, block_list& dst, uint16 skill_id, int32 heal, bool success );
```

File: src/map/mob.cpp [hunk 3]
```diff
@@ -3682,7 +3682,7 @@ int mob_summonslave(struct mob_data *md2,int *value,int amount,uint16 skill_id)
 		if (md2->state.copy_master_mode)
 			md->status.mode = md2->status.mode;
 
-		clif_skill_nodamage(&md->bl,&md->bl,skill_id,amount,1);
+		clif_skill_nodamage(&md->bl,md->bl,skill_id,amount,1);
```
Comment (Atemo) [inline src/map/mob.cpp:3685]:
```suggestion
		clif_skill_nodamage(&md->bl,md->bl,skill_id,amount,true);
```

File: src/map/pc.cpp [hunk 4]
```diff
@@ -10048,7 +10048,7 @@ bool pc_revive_item(map_session_data *sd) {
 	else
 		pc_delitem(sd, item_position, 1, 0, 1, LOG_TYPE_CONSUME);
 
-	clif_skill_nodamage(&sd->bl, &sd->bl, ALL_RESURRECTION, 4, 1);
+	clif_skill_nodamage(&sd->bl, sd->bl, ALL_RESURRECTION, 4, 1);
```
Comment (Atemo) [inline src/map/pc.cpp:10051]:
```suggestion
	clif_skill_nodamage(&sd->bl, sd->bl, ALL_RESURRECTION, 4, true);
```

File: src/map/pet.cpp [hunk 5]
```diff
@@ -2073,7 +2073,7 @@ TIMER_FUNC(pet_recovery_timer){
 	if(sd->sc.getSCE(pd->recovery->type)) {
 		//Display a heal animation?
 		//Detoxify is chosen for now.
-		clif_skill_nodamage(&pd->bl,&sd->bl,TF_DETOXIFY,1,1);
+		clif_skill_nodamage(&pd->bl,sd->bl,TF_DETOXIFY,1,1);
```
Comment (Atemo) [inline src/map/pet.cpp:2076]:
```suggestion
		clif_skill_nodamage(&pd->bl,sd->bl,TF_DETOXIFY,1,true);
```

File: src/map/pet.cpp [hunk 6]
```diff
@@ -2116,7 +2116,7 @@ TIMER_FUNC(pet_heal_timer){
 
 	pet_stop_attack(pd);
 	pet_stop_walking(pd,1);
-	clif_skill_nodamage(&pd->bl,&sd->bl,AL_HEAL,pd->s_skill->lv,1);
+	clif_skill_nodamage(&pd->bl,sd->bl,AL_HEAL,pd->s_skill->lv,1);
```
Comment (Atemo) [inline src/map/pet.cpp:2119]:
```suggestion
	clif_skill_nodamage(&pd->bl,sd->bl,AL_HEAL,pd->s_skill->lv,true);
```

File: src/map/script.cpp [hunk 7]
```diff
@@ -15343,7 +15343,7 @@ static inline void script_skill_effect( block_list& bl, uint16 skill_id, uint16
 			clif_skill_poseffect(&bl, skill_id, skill_lv, x, y, gettick());
 			break;
 		case CAST_NODAMAGE:
-			clif_skill_nodamage(&bl, &bl, skill_id, skill_lv, 1);
+			clif_skill_nodamage(&bl, bl, skill_id, skill_lv, 1);
```
Comment (Atemo) [inline src/map/script.cpp:15346]:
```suggestion
			clif_skill_nodamage(&bl, bl, skill_id, skill_lv, true);
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -6088,46 +6088,36 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
 
-	WBUFW(buf,0) = cmd;
-	WBUFW(buf,2) = skill_id;
-#if PACKETVER < 20130731
-	WBUFW(buf,4) = min(heal, INT16_MAX);
+	p.PacketType = HEADER_ZC_USE_SKILL;
+	p.SKID = skill_id;
+#if PACKETVER_MAIN_NUM >= 20130731 || PACKETVER_RE_NUM >= 20130724 || defined(PACKETVER_ZERO)
+	p.level = min(heal, INT_MAX);
 #else
```
Comment (Atemo) [inline src/map/clif.cpp:6099]:
```suggestion
	p.level = std::min( static_cast<decltype(p.level)>( heal ), std::numeric_limits<decltype(p.level)>::max() );
```

File: src/map/clif.cpp [hunk 9]
```diff
@@ -6088,46 +6088,36 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
-	WBUFL(buf,4) = min(heal, INT32_MAX);
-	offset += 2;
+	p.level = min(heal, INT16_MAX);
 #endif
```
Comment (Atemo) [inline src/map/clif.cpp:6101]:
```suggestion
```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -6088,46 +6088,36 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
-	WBUFL(buf,6+offset) = dst->id;
-	WBUFL(buf,10+offset) = src ? src->id : 0;
-	WBUFB(buf,14+offset) = success;
+	p.targetAID = dst.id;
+	p.srcAID = src != nullptr ? src->id : 0;
+	p.result = ( tick != 0 );
```
Comment (Atemo) [inline src/map/clif.cpp:6104]:
```suggestion
	p.result = success;
```

File: src/map/clif.cpp [hunk 11]
```diff
@@ -6088,46 +6088,36 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
-		clif_send(buf, packet_len(cmd), src, SELF);
+	if(src != nullptr && disguised(src)) {
+		p.srcAID = disguised_bl_id(src->id);
+		if (disguised(&dst))
+			p.targetAID = dst.id;
+		clif_send(&p, sizeof(p), src, SELF);
 	}
 
-	return success;
+	return ( tick != 0 );
```
Comment (Atemo) [inline src/map/clif.cpp:6120]:
```suggestion
	return success;
```

File: src/map/skill.cpp [hunk 12]
```diff
@@ -5287,7 +5287,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 	case TR_RHYTHMSHOOTING:
 	case HN_MEGA_SONIC_BLOW:
 	case HN_SPIRAL_PIERCE_MAX:
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5290]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 13]
```diff
@@ -5298,7 +5298,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 			sc_start(src, src, SC_E_SLASH_COUNT, 100, min( 5, 1 + sc->getSCE(SC_E_SLASH_COUNT)->val1 ), skill_get_time(skill_id, skill_lv));
 		else
 			sc_start(src, src, SC_E_SLASH_COUNT, 100, 1, skill_get_time(skill_id, skill_lv));
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5301]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 14]
```diff
@@ -5309,12 +5309,12 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 		status_change_end(src, SC_CLOAKING);
 		status_change_end(src, SC_CLOAKINGEXCEED);
 
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5312]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 15]
```diff
@@ -5309,12 +5309,12 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
...
 		skill_attack(BF_WEAPON, src, src, bl, skill_id, skill_lv, tick, flag);
 		break;
 
 	case WH_CRESCIVE_BOLT:
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5317]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 16]
```diff
@@ -5327,7 +5327,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 		if (skill_check_unit_movepos(5, src, bl->x, bl->y, 0, 1))
 			skill_blown(src, src, 1, (map_calc_dir(bl, src->x, src->y) + 4) % 8, BLOWN_NONE);
 
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5330]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 17]
```diff
@@ -5364,7 +5364,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 		break;
 
 	case TK_STORMKICK: // Taekwon kicks [Dralnu]
-		clif_skill_nodamage(src,bl,skill_id,skill_lv,1);
+		clif_skill_nodamage(src,*bl,skill_id,skill_lv,1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5367]:
```suggestion
		clif_skill_nodamage(src,*bl,skill_id,skill_lv);
```

File: src/map/skill.cpp [hunk 18]
```diff
@@ -5434,7 +5434,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 	case AG_CRIMSON_ARROW:
 		skill_area_temp[1] = bl->id;
 		if (skill_id == AG_STORM_CANNON || skill_id == AG_CRIMSON_ARROW)
-			clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+			clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5437]:
```suggestion
			clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 19]
```diff
@@ -5572,7 +5572,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 		break;
 
 	case SU_PICKYPECK:
-		clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+		clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5575]:
```suggestion
		clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/skill.cpp [hunk 20]
```diff
@@ -5768,7 +5768,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 			switch (skill_id) {
 				case NPC_VAMPIRE_GIFT:
 					if (heal > 0) {
-						clif_skill_nodamage(nullptr, src, AL_HEAL, heal, 1);
+						clif_skill_nodamage(nullptr, *src, AL_HEAL, heal, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5771]:
```suggestion
						clif_skill_nodamage(nullptr, *src, AL_HEAL, heal);
```

File: src/map/skill.cpp [hunk 21]
```diff
@@ -5806,7 +5806,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 				case DK_HACKANDSLASHER:
 				case MT_SPARK_BLASTER:
 				case HN_JUPITEL_THUNDER_STORM:
-					clif_skill_nodamage(src,bl,skill_id,skill_lv,1);
+					clif_skill_nodamage(src,*bl,skill_id,skill_lv,1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5809]:
```suggestion
					clif_skill_nodamage(src,*bl,skill_id,skill_lv);
```

File: src/map/skill.cpp [hunk 22]
```diff
@@ -5832,7 +5832,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 					// Jump to the target before attacking.
 					if (skill_check_unit_movepos(5, src, bl->x, bl->y, 0, 1))
 						skill_blown(src, src, 1, (map_calc_dir(bl, src->x, src->y) + 4) % 8, BLOWN_NONE);
-					clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);// Trigger animation on servants.
+					clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);// Trigger animation on servants.
```
Comment (AoShinRO) [inline src/map/skill.cpp:5835]:
```suggestion
					clif_skill_nodamage(src, *bl, skill_id, skill_lv);// Trigger animation on servants.
```

File: src/map/skill.cpp [hunk 23]
```diff
@@ -5848,7 +5848,7 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 					// Move the player 1 cell near the target, between the target and the player
 					if (skill_check_unit_movepos(5, src, bl->x + dirx[dir], bl->y + diry[dir], 0, 1))
 						clif_blown(src);
-					clif_skill_nodamage(src, bl, skill_id, skill_lv, 1);
+					clif_skill_nodamage(src, *bl, skill_id, skill_lv, 1);
```
Comment (AoShinRO) [inline src/map/skill.cpp:5851]:
```suggestion
					clif_skill_nodamage(src, *bl, skill_id, skill_lv);
```

File: src/map/clif.cpp [hunk 24]
```diff
@@ -6088,43 +6088,29 @@ int clif_skill_damage2(struct block_list *src,struct block_list *dst,t_tick tick
...
-	offset += 2;
-#endif
-	WBUFL(buf,6+offset) = dst->id;
-	WBUFL(buf,10+offset) = src ? src->id : 0;
-	WBUFB(buf,14+offset) = success;
+	p.PacketType = HEADER_ZC_USE_SKILL;
+	p.SKID = skill_id;
+	p.level = std::min( static_cast<decltype(p.level)>( heal ), std::numeric_limits<decltype(p.level)>::max() );
+	p.targetAID = dst.id;
+	p.srcAID = src != nullptr ? src->id : 0;
```
Comment (AoShinRO) [inline src/map/clif.cpp:6099]:
```suggestion
	p.srcAID = 0;
	if(src != nullptr)
		p.srcAID = src->id;
```

**Comment by Daegaladh** [comment]
Please, replace the bool on all the other instances

**Comment by AoShinRO** [comment]
> Please, replace the bool on all the other instances

yep, i need to open the branch again in my desktop but i'll do

File: src/map/skill.cpp [hunk 25]
```diff
@@ -10821,22 +10821,22 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
 	case ALL_CATCRY:
 	case ALL_DREAM_SUMMERNIGHT:
-		clif_skill_nodamage(src,bl,skill_id,skill_lv,1);
+		clif_skill_nodamage(src,*bl,skill_id,skill_lv);
 		break;
 	case ALL_BUYING_STORE:
 		if( sd )
 		{// players only, skill allows 5 buying slots
-			clif_skill_nodamage(src, bl, skill_id, skill_lv, buyingstore_setup(sd, MAX_BUYINGSTORE_SLOTS) ? 0 : 1);
+			clif_skill_nodamage(src, *bl, skill_id, skill_lv, !static_cast<bool>(buyingstore_setup(sd, MAX_BUYINGSTORE_SLOTS)));
```
Comment (AoShinRO) [inline src/map/skill.cpp:10835]: anyone can confirm  this call?

File: src/map/skill.cpp [hunk 26]
```diff
@@ -8555,27 +8555,27 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			break;
 		}
 		if (i) status_heal(src, 0, i, 3);
-		clif_skill_nodamage(src,bl,skill_id,skill_lv,i?1:0);
+		clif_skill_nodamage(src,*bl,skill_id,skill_lv,static_cast<bool>(i));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:8558]:
```suggestion
		clif_skill_nodamage(src,*bl,skill_id,skill_lv,i != 0);
```

File: src/map/skill.cpp [hunk 27]
```diff
@@ -10830,22 +10830,22 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
 	case ALL_CATCRY:
 	case ALL_DREAM_SUMMERNIGHT:
-		clif_skill_nodamage(src,bl,skill_id,skill_lv,1);
+		clif_skill_nodamage(src,*bl,skill_id,skill_lv);
 		break;
 	case ALL_BUYING_STORE:
 		if( sd )
 		{// players only, skill allows 5 buying slots
-			clif_skill_nodamage(src, bl, skill_id, skill_lv, buyingstore_setup(sd, MAX_BUYINGSTORE_SLOTS) ? 0 : 1);
+			clif_skill_nodamage(src, *bl, skill_id, skill_lv, !static_cast<bool>(buyingstore_setup(sd, MAX_BUYINGSTORE_SLOTS)));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10844]:
```suggestion
			clif_skill_nodamage(src, *bl, skill_id, skill_lv, buyingstore_setup(sd, MAX_BUYINGSTORE_SLOTS) == 0);
```

File: src/map/skill.cpp [hunk 28]
```diff
@@ -11435,7 +11435,7 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			}
 			heal = dstsd->status.max_hp * hp / 100;
 			status_heal(bl,heal,0,2);
-			clif_skill_nodamage(src, bl, skill_id, skill_lv, heal);
+			clif_skill_nodamage(src, *bl, skill_id, skill_lv, static_cast<bool>(heal));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:11438]:
```suggestion
			clif_skill_nodamage(src, *bl, skill_id, skill_lv, heal != 0);
```

File: src/map/skill.cpp [hunk 29]
```diff
@@ -11628,7 +11628,7 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			}
 			if (i)
 				status_percent_heal(src, 0, i);
-			clif_skill_nodamage(src, bl, skill_id, skill_lv, i ? 1:0);
+			clif_skill_nodamage(src, *bl, skill_id, skill_lv, static_cast<bool>(i));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:11631]:
```suggestion
			clif_skill_nodamage(src, *bl, skill_id, skill_lv, i != 0);
```

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**